### PR TITLE
Resolve `min` as the earliest compatible Julia version (compatible with the user's project)

### PR DESCRIPTION
**Pull request by omus**
Introduces a special version ~"MIN"~ "min" which when passed in as the `version` allows the user to test against the earliest version of Julia compatible with their project. In order to determine this version the project file is read.

~If we want we can choose another name for this special version but I would like to keep the name short as this usually appears in the CI check name which can already be quite long.~

Update: We've decided on `"min"` as the special version name

### CODE REVIEW

**Comment by omus** [comment]
Did some basic local testing but need to add real tests before this is ready for review

**Comment by DilumAluthge** [comment]
Does this work for all `julia = "..."` compat entries, or just some? Can we document which ones it works for?

File: __tests__/installer.test.ts
Comment (DilumAluthge) [inline __tests__/installer.test.ts:1]: Can you add tests for:
1. Multiple entries separated by comma, e.g. `~1.7, ~1.8, ~1.9`
2. Hyphen ranges, e.g. `1.7 - 1.9`

Comment (DilumAluthge) [reply __tests__/installer.test.ts:1]: And probably should also add tests for inequalities:
1. ` < 1.9.0`
2. `>= 1.6.0`

Comment (DilumAluthge) [reply __tests__/installer.test.ts:1]: Probably also want to test the "multiple comma-separated entries" with a couple different permutations.

So e.g. we should test `julia = "~1.7, ~1.8, ~1.9`.

But we should also test `julia = "~1.8, ~1.7, ~1.9` and make sure it gives the same result.

Comment (omus) [reply __tests__/installer.test.ts:1]: I'll add tests for all of these but based upon what I've seen in https://github.com/npm/node-semver#ranges all of these should work

Comment (omus) [reply __tests__/installer.test.ts:1]: Done

**Comment by DilumAluthge** [comment]
We should make sure that the case-convention in this PR is consistent with that in https://github.com/julia-actions/setup-julia/pull/234#issuecomment-2179549945

File: action.yml [hunk 1]
```diff
@@ -17,6 +17,10 @@ inputs:
     description: 'Display InteractiveUtils.versioninfo() after installing'
     required: false
     default: 'false'
+  project:
```
Comment (DilumAluthge) [inline action.yml:20]: Should this be called `julia-project` instead?

File: src/installer.ts [hunk 2]
```diff
@@ -76,15 +77,105 @@ export async function getJuliaVersions(versionInfo): Promise<string[]> {
...
+    }
+
+    if (!compatRange) {
+        throw new Error(`Invalid version range found in Julia compat: ${compatRange}`)
+    }
+
+    return compatRange
+}
+
+export function getJuliaVersion(availableReleases: string[], versionInput: string, includePrerelease: boolean = false, juliaCompatRange: string = ""): string {
```
Comment (DilumAluthge) [inline src/installer.ts:161]:
```suggestion
export function getJuliaVersion(availableReleases: string[], versionInput: string, includePrerelease: boolean = false, juliaCompatRange: string = ""): string {
    // Note: `juliaCompatRange` is ignored unless `versionInput` is `MIN`
```

File: src/installer.ts [hunk 3]
```diff
@@ -76,15 +77,105 @@ export async function getJuliaVersions(versionInfo): Promise<string[]> {
...
+        version = semver.minSatisfying(availableReleases, juliaCompatRange, {includePrerelease})
+    } else {
+        // Use the highest available version that matches versionInput
+        version = semver.maxSatisfying(availableReleases, versionInput, {includePrerelease})
     }
 
-    // Use the highest available version that matches versionInput
-    let version = semver.maxSatisfying(availableReleases, versionInput, {includePrerelease})
-    if (version == null) {
+    if (!version) {
```
Comment (DilumAluthge) [inline src/installer.ts:178]: What's the difference between `(!version)` (this PR) and `if (version == null)` (on master)?

File: src/installer.ts [hunk 4]
```diff
@@ -76,15 +77,105 @@ export async function getJuliaVersions(versionInfo): Promise<string[]> {
     return versions
 }
 
-export function getJuliaVersion(availableReleases: string[], versionInput: string, includePrerelease: boolean = false): string {
+export function getProjectFile(projectInput: string = ""): string {
+    let projectFile: string = ""
+
+    // Default value for projectInput
+    if (!projectInput) {
+        projectInput = process.env.JULIA_PROJECT || "."
```
Comment (DilumAluthge) [inline src/installer.ts:85]: What will happen here if:
1. `JULIA_PROJECT` is defined, but it is set to the empty string.
2. `JULIA_PROJECT` is defined, and it is nonempty, but it consists only of whitespace.

File: src/installer.ts [hunk 5]
```diff
@@ -76,15 +77,105 @@ export async function getJuliaVersions(versionInfo): Promise<string[]> {
     return versions
 }
 
-export function getJuliaVersion(availableReleases: string[], versionInput: string, includePrerelease: boolean = false): string {
+export function getProjectFile(projectInput: string = ""): string {
```
Comment (DilumAluthge) [inline src/installer.ts:80]: Might be good to rename this something like `getProjectFilename`? To emphasize that it's only returning the filename, and not the file contents?

**Review by DilumAluthge — DISMISSED** [review]

**Comment by DilumAluthge** [comment]
@omus Looks like there are some merge conflicts.

File: action.yml (hunk 1 above)
Comment (omus) [reply action.yml:20]: I mirrored what other Julia GHAs did like `julia-actions/julia-runtest`: https://github.com/julia-actions/julia-runtest/blob/cee4ae9b6957e3bb443e499c39d5c29c752f80b7/action.yml#L29-L31

File: src/installer.ts (hunk 3 above)
Comment (omus) [reply src/installer.ts:178]: So the issue with the original code is that `null == undefined` is also true so the original statement is kind of misleading. I switched things to be based on "truthy"-ness so now we thrown an exception on any value that is coerced into a `false` such as `null`, `undefined`, or `""`.

File: src/installer.ts (hunk 4 above)
Comment (omus) [reply src/installer.ts:85]:
```js
> JULIA_PROJECT = ""; JULIA_PROJECT || "."
'.'
> JULIA_PROJECT = " "; JULIA_PROJECT || "."
' '
```

File: action.yml (hunk 1 above)
Comment (omus) [reply action.yml:20]: I think I'll probably rename some of the internal functions to use the "Julia project" terminology though to be clearer

**Comment by DilumAluthge** [comment]
> With #234 merged I think it makes sense to make this lowercase `min`

Seems reasonable to me.

**Comment by IanButterworth** [comment]
@omus do you agree with that change? I could take this PR over?

File: action.yml (hunk 1 above)
Comment (omus) [reply action.yml:20]: I've added a comment internally

File: src/installer.ts (hunk 4 above)
Comment (omus) [reply src/installer.ts:91]: I've added additional tests around using `JULIA_PROJECT`

File: src/installer.ts (hunk 5 above)
Comment (omus) [reply src/installer.ts]: Function was renamed to `getProjectFilePath`

**Review by IanButterworth — DISMISSED** [review]

**Review by IanButterworth — APPROVED** [review]

**Comment by giordano** [comment]
This should be reverted as soon as possible: #278

**Comment by IanButterworth** [comment]
Should be fixed. Sorry. I forgot to update the tag to the tip of the release branch after missing adding that index file.
(Our detailed dev docs need some additional check...)

**Pull request merged**